File: problems-adapt/0000-0100/0099_rolling-window-maxima/solution_monotonic_deque.py
from collections import deque


class Solution:
    def rollingWindowMaxima(self, nums: list[int], k: int) -> list[int]:
        dq = deque()  # indices, values decreasing
        result = []
        for i, value in enumerate(nums):
            while dq and nums[dq[-1]] <= value:
                dq.pop()
            dq.append(i)
            if dq[0] <= i - k:
                dq.popleft()
            if i >= k - 1:
                result.append(nums[dq[0]])
        return result

File: problems-originals/0201-0300/0239_sliding-window-maximum/solution_monotonic_deque.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number[]}
 */
var maxSlidingWindow = function (nums, k) {
    const dq = []; // indices, values decreasing
    const result = [];
    for (let i = 0; i < nums.length; i++) {
        const value = nums[i];
        while (dq.length && nums[dq[dq.length - 1]] <= value) dq.pop();
        dq.push(i);
        if (dq[0] <= i - k) dq.shift();
        if (i >= k - 1) result.push(nums[dq[0]]);
    }
    return result;
};
